utils/check_copies.py

```diff
@@ -125,13 +126,213 @@
 transformers_module = direct_transformers_import(TRANSFORMERS_PATH)
 
 
+def _is_definition_header_ending_line(line: str) -> bool:
+    # Helper function. Returns `True` if `line` is the end parenthesis of a class/function definition
+    return re.search(r"^\s*\)(\s*->.*:|:)\s*$", line) is not None
+
+
 def _should_continue(line: str, indent: str) -> bool:
     # Helper function. Returns `True` if `line` is empty, starts with the `indent` or is the end parenthesis of a
-    # function definition
-    return line.startswith(indent) or len(line.strip()) == 0 or re.search(r"^\s*\)(\s*->.*:|:)\s*$", line) is not None
+    # class/function definition
+    return line.startswith(indent) or len(line.strip()) == 0 or _is_definition_header_ending_line(line)
+
+
+def _sanity_check_splits(splits_1, splits_2, is_class):
+    """Check the two (inner) block structures of the corresponding code block given by `split_code_into_blocks` match.
+
+    For the case of `class`, they must be of one of the following 3 cases:
+
+        - a single block without name:
+
+            class foo:
+                a = 1
+
+        - a consecutive sequence of (1 or more) blocks with name
+
+            class foo:
+
+                def f(x):
+                    return x
+
+        - a block without name, followed by a consecutive sequence of (1 or more) blocks with name
+
+            class foo:
+                a = 1
+
+                def f(x):
+                    return x
+
+                def g(x):
+                    return None
+
+    The 2 code snippets that give `splits_1` and `splits_2` have to be in the same case to pass this check, but the
+    number of blocks with name in the consecutive sequence is not taken into account.
+
+    For the case of `function or method`, we don't require it to be in one of the above 3 cases. However, the structure
+    of`splits_1` and `splits_2` have to match exactly. In particular, the number of blocks with name in a consecutive
+    sequence is taken into account.
```

Comment:
@ArthurZucker let me know if this is easier to understand.

Reply:
LGTM thanks for updating sir!